File: TigerEx/deposits_withdrawals_and_payments/crypto_deposit_system/wallet_service.ts
/**
 * TigerEx Crypto Wallet Service
 * Deposit address generation and transaction handling
 */

const SUPPORTED_ASSETS = ['BTC', 'ETH', 'USDT', 'USDC', 'BNB', 'SOL', 'XRP', 'ADA'];

class WalletService {
  /**
   * Generate deposit address for user
   */
  async generateDepositAddress(userId: string, currency: string, network: string) {
    // Generate deterministic address from user ID
    const address = this.deriveAddress(currency, userId, network);
    
    return {
      address,
      currency,
      network,
      qrCode: `/${currency}:${address}`,
      memo: network === 'sol' ? 'MEMO' : undefined
    };
  }

  /**
   * Process incoming deposit
   */
  async processDeposit(txHash: string, data: any): Promise<any> {
    // Verify transaction
    const txData = await this.verifyOnChain(txHash);
    
    if (!txData.confirmed) {
      throw new Error('Transaction not confirmed');
    }
    
    // Screen against sanctions
    if (await this.screenAddress(txData.from)) {
      throw new Error('Address blocked');
    }
    
    // Create transaction record
    const tx = {
      id: crypto.randomUUID(),
      type: 'deposit',
      currency: data.currency,
      amount: txData.amount,
      txHash,
      status: 'completed',
      userId: data.userId,
      confirmations: txData.confirmations,
      createdAt: new Date()
    };
    
    return tx;
  }

  /**
   * Process withdrawal request
   */
  async processWithdrawal(request: any): Promise<any> {
    // Check balance
    const balance = await this.getBalance(request.userId, request.currency);
    if (balance < request.amount) {
      throw new Error('Insufficient balance');
    }
    
    // Validate address
    if (!this.validateAddress(request.address, request.currency)) {
      throw new Error('Invalid address');
    }
    
    // Create pending transaction
    const tx = {
      id: crypto.randomUUID(),
      type: 'withdrawal',
      currency: request.currency,
      amount: request.amount,
      address: request.address,
      status: 'pending',
      fee: await this.calculateFee(request.currency),
      createdAt: new Date()
    };
    
    return tx;
  }

  /**
   * Get user wallet balance
   */
  async getBalance(userId: string, currency: string): Promise<number> {
    // Simplified - would query database
    return 10000;
  }

  /**
   * Validate cryptocurrency address
   */
  validateAddress(address: string, currency: string): boolean {
    switch (currency) {
      case 'BTC':
        return /^(bc1|[13])[a-zA-HJ-NP-Z0-9]{25,62}$/.test(address);
      case 'ETH':
      case 'USDT':
      case 'USDC':
        return /^0x[a-fA-F0-9]{40}$/.test(address);
      case 'SOL':
        return /^[1-9A-HJ-NP-Za-km-z]{32,44}$/.test(address);
      default:
        return address.length > 20;
    }
  }

  /**
   * Derive deposit address
   */
  private deriveAddress(currency: string, userId: string, network: string): string {
    // Simplified HD wallet derivation
    return `0x${crypto.subtle.digest('SHA-256', userId + currency).then(h => 
      Array.from(new Uint8Array(h)).map(b => b.toString(16).padStart(2, '0')).join('').slice(0, 40)
    );
  }

  /**
   * Verify transaction on blockchain
   */
  private async verifyOnChain(txHash: string): Promise<any> {
    return {
      from: '0xsender...',
      to: '0xreceiver...',
      amount: 1000,
      confirmations: 12,
      confirmed: true
    };
  }

  /**
   * Screen address
   */
  private async screenAddress(address: string): Promise<boolean> {
    // Check sanctions list
    return false;
  }

  /**
   * Calculate network fee
   */
  private async calculateFee(currency: string): Promise<number> {
    const fees: Record<string, number> = {
      'BTC': 0.0005,
      'ETH': 0.005,
      'USDT': 1,
      'USDC': 1
    };
    return fees[currency] || 0.01;
  }
}

// ============================================================================
// P2P Trading Platform
// ============================================================================

class PPTradingService {
  /**
   * Create P2P advertisement
   */
  async createAd(params: any): Promise<any> {
    const ad = {
      id: crypto.randomUUID(),
      userId: params.userId,
      type: params.type,  // BUY or SELL
      currency: params.currency,
      fiatCurrency: params.fiatCurrency,
      priceType: 'fixed' | '浮动',
      priceOffset: params.priceOffset,
      limits: params.limits,
      paymentMethods: params.paymentMethods,
      terms: params.terms,
      status: 'active',
      createdAt: new Date()
    };
    
    return ad;
  }

  /**
   * Create P2P order
   */
  async createOrder(adId: string, buyerId: string, amount: number): Promise<any> {
    const order = {
      id: crypto.randomUUID(),
      adId,
      buyerId,
      amount,
      status: 'pending',
      createdAt: new Date()
    };
    
    return order;
  }

  /**
   * Mark payment made
   */
  async markPayment(orderId: string, buyerId: string): Promise<any> {
    return {
      status: 'paid',
      paidAt: new Date()
    };
  }

  /**
   * Confirm release (seller releases crypto)
   */
  async confirmRelease(orderId: string, sellerId: string): Promise<any> {
    return {
      status: 'completed',
      completedAt: new Date()
    };
  }

  /**
   * Cancel order
   */
  async cancelOrder(orderId: string, userId: string, reason: string): Promise<any> {
    return {
      status: 'cancelled',
      reason,
      cancelledAt: new Date()
    };
  }

  /**
   * Dispute order
   */
  async openDispute(orderId: string, userId: string, reason: string): Promise<any> {
    return {
      status: 'disputed',
      reason,
      disputeOpenedAt: new Date(),
      assignedArbitrator: 'system'
    };
  }

  /**
   * Resolve dispute
   */
  async resolveDispute(orderId: string, resolution: 'release' | 'refund'): Promise<any> {
    return {
      status: resolution === 'release' ? 'completed' : 'cancelled',
      resolvedAt: new Date(),
      resolution
    };
  }
}

export { WalletService, PPTradingService, SUPPORTED_ASSETS };

File: TigerEx/deposits_withdrawals_and_payments/crypto_deposit_system/wallet_service.go
package wallet

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"regexp"
	"strings"
)

// Supported assets
var SUPPORTED_ASSETS = []string{"BTC", "ETH", "USDT", "USDC", "BNB", "SOL", "XRP", "ADA"}

// WalletService handles crypto wallets
type WalletService struct {
	db interface{}
}

// NewWalletService creates new wallet service
func NewWalletService(db interface{}) *WalletService {
	return &WalletService{db: db}
}

// DepositAddress represents a deposit address
type DepositAddress struct {
	Address  string `json:"address"`
	Currency string `json:"currency"`
	Network  string `json:"network"`
	QRCode   string `json:"qrCode"`
	Memo     string `json:"memo,omitempty"`
}

// GenerateDepositAddress creates a new deposit address
func (s *WalletService) GenerateDepositAddress(userID, currency, network string) (*DepositAddress, error) {
	address := s.deriveAddress(currency, userID, network)
	
	depositAddr := &DepositAddress{
		Address:  address,
		Currency: currency,
		Network:  network,
		QRCode:   fmt.Sprintf("/%s:%s", currency, address),
	}
	
	if network == "sol" {
		depositAddr.Memo = "MEMO"
	}
	
	return depositAddr, nil
}

// TxData represents transaction data
type TxData struct {
	ID          string  `json:"id"`
	Type       string  `json:"type"`
	Currency   string  `json:"currency"`
	Amount     float64 `json:"amount"`
	TxHash     string  `json:"txHash"`
	Status     string  `json:"status"`
	UserID     string  `json:"userId"`
	Confirmations int   `json:"confirmations"`
	CreatedAt   int64   `json:"createdAt"`
}

// ProcessDeposit processes an incoming deposit
func (s *WalletService) ProcessDeposit(txHash string, data interface{}) (*TxData, error) {
	txData, err := s.verifyOnChain(txHash)
	if err != nil {
		return nil, err
	}
	
	if !txData.Confirmed {
		return nil, errors.New("transaction not confirmed")
	}
	
	// Screen against sanctions
	if s.screenAddress(txData.From) {
		return nil, errors.New("address blocked")
	}
	
	// Would parse data input in production
	inputMap := data.(map[string]interface{})
	
	tx := &TxData{
		ID:          generateUUID(),
		Type:        "deposit",
		Currency:    inputMap["currency"].(string),
		Amount:      txData.Amount,
		TxHash:      txHash,
		Status:      "completed",
		UserID:      inputMap["userId"].(string),
		Confirmations: txData.Confirmations,
		CreatedAt:   nowMillis(),
	}
	
	return tx, nil
}

// WithdrawalRequest represents a withdrawal request
type WithdrawalRequest struct {
	UserID   string  `json:"userId"`
	Address string  `json:"address"`
	Amount   float64 `json:"amount"`
	Currency string `json:"currency"`
}

// ProcessWithdrawal processes a withdrawal request
func (s *WalletService) ProcessWithdrawal(request *WithdrawalRequest) (*TxData, error) {
	balance, err := s.GetBalance(request.UserID, request.Currency)
	if err != nil {
		return nil, err
	}
	
	if balance < request.Amount {
		return nil, errors.New("insufficient balance")
	}
	
	if !s.ValidateAddress(request.Address, request.Currency) {
		return nil, errors.New("invalid address")
	}
	
	fee, err := s.calculateFee(request.Currency)
	if err != nil {
		return nil, err
	}
	
	tx := &TxData{
		ID:      generateUUID(),
		Type:    "withdrawal",
		Amount: request.Amount,
		Status: "pending",
		CreatedAt: nowMillis(),
	}
	
	return tx, nil
}

// GetBalance returns wallet balance
func (s *WalletService) GetBalance(userID, currency string) (float64, error) {
	// Simplified - would query database
	return 10000.0, nil
}

// ValidateAddress validates a cryptocurrency address
func (s *WalletService) ValidateAddress(address, currency string) bool {
	switch currency {
	case "BTC":
		return regexp.MustCompile(`^(bc1|[13])[a-zA-HJ-NP-Z0-9]{25,62}$`).MatchString(address)
	case "ETH", "USDT", "USDC":
		return regexp.MustCompile(`^0x[a-fA-F0-9]{40}$`).MatchString(address)
	case "SOL":
		return regexp.MustCompile(`^[1-9A-HJ-NP-Za-km-z]{32,44}$`).MatchString(address)
	default:
		return len(address) > 20
	}
}

func (s *WalletService) deriveAddress(currency, userID, network string) string {
	hasher := sha256.New()
	hasher.Write([]byte(userID + currency))
	hash := hasher.Sum(nil)
	return "0x" + hex.EncodeToString(hash)[:40]
}

func (s *WalletService) verifyOnChain(txHash string) (*BlockchainTX, error) {
	return &BlockchainTX{
		From:         "0xsender...",
		To:           "0xreceiver...",
		Amount:       1000,
		Confirmed:    true,
		Confirmations: 12,
	}, nil
}

func (s *WalletService) screenAddress(address string) bool {
	// Check sanctions list - simplified
	return false
}

func (s *WalletService) calculateFee(currency string) (float64, error) {
	fees := map[string]float64{
		"BTC":  0.0005,
		"ETH":  0.005,
		"USDT": 1,
		"USDC": 1,
	}
	
	if fee, ok := fees[currency]; ok {
		return fee, nil
	}
	return 0.01, nil
}

// BlockchainTX verified transaction
type BlockchainTX struct {
	From         string
	To           string
	Amount      float64
	Confirmations int
	Confirmed   bool
}

// P2P Trading Service
type P2PTradingService struct {
	db interface{}
}

// P2PAd represents a P2P advertisement
type P2PAd struct {
	ID             string   `json:"id"`
	UserID        string   `json:"userId"`
	Type          string   `json:"type"`
	Currency      string   `json:"currency"`
	FiatCurrency  string   `json:"fiatCurrency"`
	PriceType     string   `json:"priceType"`
	PriceOffset   float64  `json:"priceOffset"`
	Limits       []int    `json:"limits"`
	PaymentMethods []string `json:"paymentMethods"`
	Terms        string   `json:"terms"`
	Status       string   `json:"status"`
	CreatedAt     int64    `json:"createdAt"`
}

// CreateAd creates a P2P advertisement
func (s *P2PTradingService) CreateAd(params *P2PAd) (*P2PAd, error) {
	ad := &P2PAd{
		ID:             generateUUID(),
		UserID:         params.UserID,
		Type:           params.Type,
		Currency:       params.Currency,
		FiatCurrency:  params.FiatCurrency,
		PriceType:     "fixed",
		PriceOffset:   params.PriceOffset,
		Limits:        params.Limits,
		PaymentMethods: params.PaymentMethods,
		Terms:         params.Terms,
		Status:        "active",
		CreatedAt:     nowMillis(),
	}
	
	return ad, nil
}

// P2POrder represents a P2P order
type P2POrder struct {
	ID        string `json:"id"`
	AdID     string `json:"adId"`
	BuyerID  string `json:"buyerId"`
	Amount  float64 `json:"amount"`
	Status  string `json:"status"`
	CreateAt int64 `json:"createdAt"`
}

// CreateOrder creates a P2P order
func (s *P2PTradingService) CreateOrder(adID, buyerID string, amount float64) (*P2POrder, error) {
	return &P2POrder{
		ID:        generateUUID(),
		AdID:     adID,
		BuyerID:  buyerID,
		Amount:  amount,
		Status:  "pending",
		CreateAt: nowMillis(),
	}, nil
}

// MarkPayment marks payment made
func (s *P2PTradingService) MarkPayment(orderID, buyerID string) error {
	return nil
}

// ConfirmRelease confirms crypto release
func (s *P2PTradingService) ConfirmRelease(orderID, sellerID string) error {
	return nil
}

// CancelOrder cancels a P2P order
func (s *P2PTradingService) CancelOrder(orderID, userID, reason string) error {
	return nil
}

// OpenDispute opens a dispute
func (s *P2PTradingService) OpenDispute(orderID, userID, reason string) error {
	return nil
}

// ResolveDispute resolves a dispute
func (s *P2PTradingService) ResolveDispute(orderID string, resolution string) error {
	return nil
}

// Helpers
func generateUUID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return strings.ReplaceAll(hex.EncodeToString(b), "-", "")
}

func nowMillis() int64 {
	return 0 // Simplified - would use time.Now().UnixMilli()
}